src/algebra/squarefree.lean

```diff
@@ -204,4 +204,71 @@ lemma sum_divisors_filter_squarefree {n : ℕ} (h0 : n ≠ 0)
 by rw [finset.sum_eq_multiset_sum, divisors_filter_squarefree h0, multiset.map_map,
     finset.sum_eq_multiset_sum]
 
+lemma sq_mul_squarefree {n : ℕ} (hn : 0 < n) :
+  ∃ a b : ℕ, 0 < a ∧ 0 < b ∧ b ^ 2 * a = n ∧ squarefree a :=
+begin
+  let S := {s ∈ finset.range (n + 1) | s ∣ n ∧ ∃ x, s = x ^ 2 },
+  have hSne : S.nonempty,
+  { use 1,
+    have h1 : 0 < n ∧ ∃ (x : ℕ), 1 = x ^ 2 := ⟨hn, ⟨1, (one_pow 2).symm⟩⟩,
+    simpa [S] },
+
+  let s := finset.max' S hSne,
+  have hs : s ∈ S := finset.max'_mem S hSne,
+  simp only [finset.sep_def, S, finset.mem_filter, finset.mem_range] at hs,
+  obtain ⟨hsn1, ⟨a, hsa⟩, ⟨b, hsb⟩⟩ := hs,
+
+  have hzero_lt_a : 0 < a,
+  { rw hsa at hn,
+    rw zero_lt_iff at hn ⊢,
+    exact right_ne_zero_of_mul hn },
+  have hzero_lt_s : 0 < s,
+  { rw hsa at hn,
+    rw zero_lt_iff at hn ⊢,
+    exact left_ne_zero_of_mul hn },
+  have hzero_lt_b : 0 < b,
+  { rw [hsb, pow_two] at hzero_lt_s,
+    rw zero_lt_iff at hzero_lt_s ⊢,
+    exact left_ne_zero_of_mul hzero_lt_s },
+
+  rw hsb at hsa,
+
+  refine ⟨a, b, hzero_lt_a, hzero_lt_b, hsa.symm, _⟩,
+  intro x,
+  rintro ⟨y, hy⟩,
+  by_contradiction,
+  rw nat.is_unit_iff at h,
+
+  have hx₀ : ¬(x = 0),
+  { by_contradiction hx0,
+    rw [hy, hx0, zero_mul 0, zero_mul y] at hzero_lt_a,
+    exact nat.lt_asymm hzero_lt_a hzero_lt_a },
+
+  have hx₂ : x > 1 := nat.one_lt_iff_ne_zero_and_ne_one.mpr ⟨hx₀, h⟩,
+
+  set z := (b * x) ^ 2 with hz,
```

Comment:
You can change this to `let` and change all the instances of `rw hz` to `simp_rw [hz]` (I saw that this came up in #7274 as well).

Reply:
`simp_rw` is the only `rw` tool to force `let`s to unfold? not even something like `unfold`? (also I'm guessing it'd be `simp_rw [z]`

Reply:
`simp` would do it, too. I highly doubt `unfold` would work, I don't think `erw` would, either.

Reply:
And yes, sorry about the typo

Reply:
Also I think sometimes `convert` will pick up on equality arguments implicitly.

Reply:
I am going to keep using set in this case, because that way I can do `rw [hz, hsb, mul_pow b x 2, gt_iff_lt]` in one command and also `rw [hsa, hy, hz]` in one command. It is simpler that way in my opinion.

Reply:
You can replace `rw [hz, hsb, mul_pow b x 2, gt_iff_lt]` with `simp_rw [z, hsb, mul_pow b x 2, gt_iff_lt]` and `rw [hsa, hy, hz]` with `simp_rw [hsa, hy, z]`.

But anyway I am going to recommend different change that will render this moot.